### PR TITLE
Return error when no file is uploaded.

### PR DESCRIPTION
**Pull request by richardpringle**
connect to #75

### CODE REVIEW

**Comment by richardpringle** [comment]
@loay @raymondfeng, please review

**Comment by richardpringle** [comment]
@slnode test please

File: lib/storage-handler.js [hunk 1]
```diff
@@ -174,6 +174,10 @@ exports.upload = function(provider, req, res, options, cb) {
   };
 
   form.parse(req, function(err, _fields, _files) {
+    if (!err && Object.keys(_files).length === 0) {
```
Comment (richardpringle) [inline lib/storage-handler.js]: This is what @loay is referring to

**Comment by richardpringle** [comment]
Looks like #75 came up because the behavior changes with newer versions of node. With the old verison of node, the `Content-Type` header is checked but it seems as though this does not happen with newer versions, I am not sure why.

Will have to investigate further...

**Comment by raymondfeng** [comment]
LGTM

**Comment by richardpringle** [comment]
The issue that is causing failure on Node 0.10 and 0.12 is in [Formidable](https://github.com/felixge/node-formidable/) and takes place [here](https://github.com/felixge/node-formidable/blob/d93055d42c5be9f76676f296e91aaeaf2ca3833c/lib/incoming_form.js#L246). 

On Node 4.x, `bytesExpected` is `null` instead of `0`. 

I'm not sure if this is because we aren't setting it properly of if this is an issue with **Formidable** itself.

**Comment by richardpringle** [comment]
@loay, do you want to take this over? It's just the test cases that need to be fixed, I don't think any of the code needs to change.

**Comment by bajtos** [comment]
The `form.parse` callback functions looks pretty messy to me right now:

```js
  form.parse(req, function(err, _fields, _files) {
    if (!err && Object.keys(_files).length === 0) {
      err = new Error('No file content uploaded');
      res.status(400);
    }
    if (err) {
      console.error(err);
    }
    cb && cb(err, {files: files, fields: fields});
  });
```

I am proposing to clean it up a bit along the following lines:

```js
  form.parse(req, function(err, _fields, _files) {
    cb = cb || function() {};

    if (err) {
      console.error(err);
      return cb(err);
    }

    if (Object.keys(_files).length === 0) {
      err = new Error('No file content uploaded');
      err.statusCode = 400; // DO NOT MODIFY res.status directly!
      return cb(err);
    }

    cb(null, {files: files, fields: fields});
  });
```

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by loay** [comment]
- refactored form.parse as proposed above. 
- fixed linting issues. 
- rebased. 
To be fixed: failure with node 0.10 and 0.12

**Comment by bajtos** [comment]
LGTM as far as I am concerned. Please rebase on top of the latest master and check why CI builds are failing.

**Pull request merged**